shared/types/indexer.go

```diff
@@ -0,0 +1,239 @@
+	}
+	return result, nil
+}
+
+func (x *DefaultTxResult) Hash() ([]byte, error) {
+	bz, err := x.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	return x.HashFromBytes(bz)
+}
+
+func (x *DefaultTxResult) HashFromBytes(bz []byte) ([]byte, error) {
+	return crypto.SHA3Hash(bz), nil
+}
+
+type DefaultTxIndexer struct {
+	db kvstore.KVStore
+}
+
+func (d *DefaultTxIndexer) Close() error {
+	return d.db.Stop()
+}
+
+func NewTxIndexer(databasePath string) (TxIndexer, error) {
+	db, err := kvstore.NewKVStore(databasePath)
+	return &DefaultTxIndexer{
+		db: db,
+	}, err
+}
+
+func NewMemTxIndexer() (TxIndexer, error) {
+	return &DefaultTxIndexer{
+		db: kvstore.NewMemKVStore(),
+	}, nil
+}
+
+func (d *DefaultTxIndexer) Index(result TxResult) error {
+	bz, err := result.Bytes()
+	if err != nil {
+		return err
+	}
+	hash, err := result.HashFromBytes(bz)
+	if err != nil {
+		return err
+	}
+	hashKey, err := d.indexByHash(hash, bz)
+	if err != nil {
+		return err
+	}
+	if err := d.indexByHeightAndIndex(result.GetHeight(), result.GetIndex(), hashKey); err != nil {
+		return err
+	}
+	if err := d.indexBySender(result.GetSender(), hashKey); err != nil {
+		return err
+	}
+	if err := d.indexByRecipient(result.GetRecipient(), hashKey); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *DefaultTxIndexer) GetByHash(hash []byte) (TxResult, error) {
+	return d.get(d.hashKey(hash))
+}
+
+func (d *DefaultTxIndexer) GetByHeight(height int64) ([]TxResult, error) {
+	return d.getAll(d.heightKey(height), DefaultHeightOrdering)
+}
+
+func (d *DefaultTxIndexer) GetBySender(sender string, descending bool) ([]TxResult, error) {
+	return d.getAll(d.senderKey(sender), descending)
+}
+
+func (d *DefaultTxIndexer) GetByRecipient(rec string, descending bool) ([]TxResult, error) {
+	return d.getAll(d.recipientKey(rec), descending)
+}
+
+// kv helper functions
+
+func (d *DefaultTxIndexer) getAll(prefix []byte, descending bool) (res []TxResult, err error) {
+	hashKeys, err := d.db.GetAll(prefix, descending)
+	if err != nil {
+		return nil, err
+	}
+	for _, hashKey := range hashKeys {
+		var txResult TxResult
+		txResult, err = d.get(hashKey)
+		if err != nil {
+			return
+		}
+		res = append(res, txResult)
+	}
+	return
+}
+
+func (d *DefaultTxIndexer) get(key []byte) (TxResult, error) {
+	bz, err := d.db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	return new(DefaultTxResult).FromBytes(bz)
+}
+
+// index helper functions
+
+func (d *DefaultTxIndexer) indexByHash(hash, bz []byte) (hashKey []byte, err error) {
+	key := d.hashKey(hash)
+	return key, d.db.Put(key, bz)
+}
+
+func (d *DefaultTxIndexer) indexByHeightAndIndex(height int64, index int32, bz []byte) error {
+	return d.db.Put(d.heightAndIndexKey(height, index), bz)
+}
+
+func (d *DefaultTxIndexer) indexBySender(sender string, bz []byte) error {
+	return d.db.Put(d.senderKey(sender), bz)
+}
+
+func (d *DefaultTxIndexer) indexByRecipient(recipient string, bz []byte) error {
+	if recipient == "" {
+		return nil
```

Comment:
Should this be an error? I don't think the recipient should ever be nill.

Add a unit test for this case too

Reply:
There are many transactions that don't have a recipient. In fact, there is only one transaction type that has a recipient `send`

Reply:
Noted. Adding a comment suggestion